File: extensions/coding-workflow/skills/design-issues/scripts/check_issues.py
#!/usr/bin/env python3
"""
design-issues（③Issue 拆分）硬规则机器验证

Usage:
    python3 check_issues.py <topic_dir>

检查项：
  ①结构性：issues.md 存在 / verdict:pass / 关键章节 / 无占位符 / review-issues APPROVED
  ②引用：
    - P0/P1 issue 有 ≥2 方案对比（方案 A/B）
    - blocked_by 引用的 issue 编号都存在（无幽灵依赖）
    - P 级一致性：P0 不 blocked_by P2/P3（P0 不应依赖低优先级）

Exit code: 0 = 全过，1 = 有硬伤
"""

import os
import sys

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
_LIB_DIR = os.path.join(
    os.path.dirname(os.path.abspath(__file__)),
    os.pardir, os.pardir, "design-clarity", "scripts",
)
sys.path.insert(0, _LIB_DIR)
from _shared_check_lib import (
    CheckReport, resolve_topic_dir,
    check_file_exists, check_frontmatter_verdict, check_required_sections,
    check_no_placeholders, check_review_verdict,
    extract_p_levels, extract_blocked_by, extract_issue_ids, extract_section,
)

DELIVERABLE = "issues.md"


def main():
    topic_dir = resolve_topic_dir()
    report = CheckReport("issues")
    md_path = os.path.join(topic_dir, DELIVERABLE)

    # ① 结构性
    if not check_file_exists(report, f"{DELIVERABLE} 存在", md_path):
        report.finalize_and_exit(topic_dir)
        return
    check_frontmatter_verdict(report, md_path, "pass")
    check_required_sections(
        report, md_path, "关键章节",
        [r"地图总览|DAG|决策图", r"#\d+|[Ii]ssue"],
    )
    check_no_placeholders(report, "无占位符", md_path)
    check_review_verdict(report, topic_dir, "issues", "APPROVED")

    # ② P0/P1 issue 有 ≥2 方案对比
    p_levels = extract_p_levels(md_path)
    all_issue_ids = extract_issue_ids(md_path)
    # 分段检查每个 issue 的方案数（精确匹配 ## #N 标题，不误匹配 ### 子标题）
    content = open(md_path, encoding="utf-8").read()
    import re
    insufficient_solutions = []
    for m in re.finditer(r"^##\s+#(\d+)[^\n]*\n(.*?)(?=^##\s+#\d+|\Z)", content, re.DOTALL | re.MULTILINE):
        issue_num = m.group(1)
        body = m.group(2)
        level = p_levels.get(issue_num)
        if level in ("P0", "P1"):
            # 数「方案 A」「方案 B」「#### 方案」出现次数
            solution_count = len(re.findall(r"方案\s*[A-Z]|####\s*方案", body))
            if solution_count < 2:
                insufficient_solutions.append(f"#{issue_num}({level}): 仅 {solution_count} 方案")
    if insufficient_solutions:
        report.add_fail("P0/P1 issue ≥2 方案对比", "; ".join(insufficient_solutions))
    elif p_levels:
        report.add_pass("P0/P1 issue ≥2 方案对比", "全部 P0/P1 issue 有 ≥2 方案")
    else:
        report.add_skip("P0/P1 issue ≥2 方案对比", "无 P0/P1 issue")

    # ② blocked_by 引用的 issue 都存在（无幽灵依赖）
    blocked_by = extract_blocked_by(md_path)
    ghost_deps = []
    for issue_num, deps in blocked_by.items():
        for dep in deps:
            if dep not in all_issue_ids:
                ghost_deps.append(f"#{issue_num} blocked_by #{dep}（#{dep} 不存在）")
    if ghost_deps:
        report.add_fail("blocked_by 无幽灵依赖", "; ".join(ghost_deps))
    else:
        report.add_pass("blocked_by 无幽灵依赖", "所有 blocked_by 引用都存在")

    # ② P 级一致性：P0 不 blocked_by P2/P3
    level_order = {"P0": 0, "P1": 1, "P2": 2, "P3": 3}
    level_violations = []
    for issue_num, deps in blocked_by.items():
        my_level = p_levels.get(issue_num)
        if my_level not in level_order:
            continue
        for dep in deps:
            dep_level = p_levels.get(dep)
            if dep_level in level_order:
                # P0 不应依赖 P2/P3（低优先级不应阻塞高优先级）
                if level_order[my_level] < level_order[dep_level]:
                    level_violations.append(
                        f"#{issue_num}({my_level}) blocked_by #{dep}({dep_level})"
                    )
    if level_violations:
        report.add_fail("P 级一致性", "; ".join(level_violations) + "（高优先级不应依赖低优先级）")
    else:
        report.add_pass("P 级一致性", "P 级与 blocked_by 一致")

    report.finalize_and_exit(topic_dir)


if __name__ == "__main__":
    main()

File: extensions/coding-workflow/src/cw/checks/check-issues.ts
/**
 * full-issues（③Issue 拆分）硬规则机器验证。
 *
 * 移植自 skills/full-issues/scripts/check_issues.py。
 * 入参：topicDir（含 issues.md + changes/review-issues.md）。
 *
 * 检查项：
 *   ①结构性：交付物存在 / verdict:pass / 关键章节 / 无占位符 / review-issues APPROVED
 *   ②引用：P0/P1 issue ≥2 方案对比 / blocked_by 无幽灵依赖 / P 级一致性
 *   ③覆盖核验表（形式）：表存在、每行有 #issue 或 N/A+理由、无 ❌ 待补残留
 */

import { join } from "node:path";

import {
  checkFileExists,
  checkFrontmatterVerdict,
  checkNoPlaceholders,
  type CheckOutput,
  CheckReport,
  checkRequiredSections,
  checkReviewVerdict,
  extractBlockedBy,
  extractIssueIds,
  extractPLevels,
  extractSection,
  readText,
} from "./shared.js";

const DELIVERABLE = "issues.md";

export function runCheckIssues(topicDir: string): CheckOutput {
  const report = new CheckReport("issues");
  const mdPath = join(topicDir, DELIVERABLE);

  // ① 结构性
  if (!checkFileExists(report, `${DELIVERABLE} 存在`, mdPath)) {
    return report.toOutput({ writeReport: true, topicDir });
  }
  checkFrontmatterVerdict(report, mdPath, "pass");
  checkRequiredSections(report, mdPath, "关键章节", [
    "地图总览|DAG|决策图",
    "#\\d+|[Ii]ssue",
  ]);
  checkNoPlaceholders(report, "无占位符", mdPath);
  checkReviewVerdict(report, topicDir, "issues", "APPROVED");

  // ② P0/P1 issue 有 ≥2 方案对比
  const pLevels = extractPLevels(mdPath);
  const content = readText(mdPath);
  const insufficientSolutions: string[] = [];
  // 精确匹配 ## #N 或 ### #N 标题分段（\Z → (?![\s\S])；DOTALL|MULTILINE → gs flag）
  // B3: 正则 ^#{2,3}\s+#(\d+) 同时匹配 ## #N 与 ### #N（与 ISSUE_HEADING_RE 对齐）
  const issueSegmentRe = /^#{2,3}\s+#(\d+)[^\n]*\n([\s\S]*?)(?=^#{2,3}\s+#\d+|(?![\s\S]))/gm;
  for (const m of content.matchAll(issueSegmentRe)) {
    const issueNum = m[1]!;
    const body = m[2]!;
    const level = pLevels[issueNum];
    if (level === "P0" || level === "P1") {
      const solutionCount = (body.match(/方案\s*[A-Z]|####\s*方案/g) ?? []).length;
      const MIN_SOLUTIONS_FOR_P0P1 = 2; // P0/P1 issue 至少 2 方案对比
      if (solutionCount < MIN_SOLUTIONS_FOR_P0P1) {
        insufficientSolutions.push(`#${issueNum}(${level}): 仅 ${solutionCount} 方案`);
      }
    }
  }
  if (insufficientSolutions.length > 0) {
    report.addFail("P0/P1 issue ≥2 方案对比", insufficientSolutions.join("; "));
  } else if (Object.keys(pLevels).length > 0) {
    report.addPass("P0/P1 issue ≥2 方案对比", "全部 P0/P1 issue 有 ≥2 方案");
  } else {
    report.addSkip("P0/P1 issue ≥2 方案对比", "无 P0/P1 issue");
  }

  // ② blocked_by 引用的 issue 都存在（无幽灵依赖）
  const allIssueIds = new Set(extractIssueIds(mdPath));
  const blockedBy = extractBlockedBy(mdPath);
  const ghostDeps: string[] = [];
  for (const [issueNum, deps] of Object.entries(blockedBy)) {
    for (const dep of deps) {
      if (!allIssueIds.has(dep)) {
        ghostDeps.push(`#${issueNum} blocked_by #${dep}（#${dep} 不存在）`);
      }
    }
  }
  if (ghostDeps.length > 0) {
    report.addFail("blocked_by 无幽灵依赖", ghostDeps.join("; "));
  } else {
    report.addPass("blocked_by 无幽灵依赖", "所有 blocked_by 引用都存在");
  }

  // ② P 级一致性：P0 不 blocked_by P2/P3（高优先级不应依赖低优先级）
  const levelOrder: Record<string, number> = { P0: 0, P1: 1, P2: 2, P3: 3 };
  const levelViolations: string[] = [];
  for (const [issueNum, deps] of Object.entries(blockedBy)) {
    const myLevel = pLevels[issueNum];
    if (myLevel === undefined || !(myLevel in levelOrder)) continue;
    for (const dep of deps) {
      const depLevel = pLevels[dep];
      if (depLevel !== undefined && depLevel in levelOrder) {
        if (levelOrder[myLevel]! < levelOrder[depLevel]!) {
          levelViolations.push(`#${issueNum}(${myLevel}) blocked_by #${dep}(${depLevel})`);
        }
      }
    }
  }
  if (levelViolations.length > 0) {
    report.addFail(
      "P 级一致性",
      `${levelViolations.join("; ")}（高优先级不应依赖低优先级）`,
    );
  } else {
    report.addPass("P 级一致性", "P 级与 blocked_by 一致");
  }

  // ③ 覆盖核验表形式检查
  // 真 issue 定义 = 出现在 ## #N / ### #N 标题里的编号（非表格引用）
  const realIssueIds = new Set(extractRealIssueIds(content));
  checkCoverageTable(report, mdPath, realIssueIds);

  return report.toOutput({ writeReport: true, topicDir });
}

/**
 * 提取「真 issue 定义」编号：`^#{2,3}\s+#(\d+)` 标题里的 #N。
 * 区分表格引用（行中 #N）和标题定义（行首 ## #N）。
 */
export function extractRealIssueIds(content: string): string[] {
  const re = /^#{2,3}\s+#(\d+)/gm;
  const out: string[] = [];
  for (const m of content.matchAll(re)) {
    if (m[1]) out.push(m[1]);
  }
  return out;
}

/**
 * 覆盖核验表形式检查。
 *
 * 只验形式——查不了漏行/虚标/弱理由（实质靠 Step2 独立重建）。
 * realIssueIds = 从 ## #N / ### #N 章节标题提取的「真 issue 定义」编号集合。
 */
function checkCoverageTable(
  report: CheckReport,
  mdPath: string,
  realIssueIds: Set<string>,
): void {
  const section = extractSection(mdPath, "上游覆盖核验");
  if (!section) {
    report.addFail("覆盖核验表存在", "缺「上游覆盖核验」章节（MANDATORY）");
    return;
  }

  const lines = section.split(/\r?\n/).filter((ln) => ln.trim().startsWith("|"));
  let issueCol: number | null = null;
  let statusCol: number | null = null;
  let reasonCol: number | null = null;
  const dataRows: string[][] = [];

  for (const s of lines) {
    // 分隔行（纯 -/: 组合）
    if (/^\|[\s:|-]+\|?$/.test(s.trim())) continue;
    const cells = s.trim().replace(/^\|/, "").replace(/\|$/, "").split("|").map((c) => c.trim());
    if (cells.length > 0 && cells[0] === "上游元素") {
      // 表头
      for (let idx = 0; idx < cells.length; idx++) {
        const c = cells[idx]!;
        const low = c.toLowerCase();
        if (issueCol === null && (low.includes("issue") || c.includes("对应"))) issueCol = idx;
        if (statusCol === null && c === "状态") statusCol = idx;
        if (reasonCol === null && (c.includes("理由") || (low.includes("n/a") && c.includes("理由")))) {
          reasonCol = idx;
        }
      }
      continue;
    }
    dataRows.push(cells);
  }

  if (dataRows.length === 0) {
    report.addFail("覆盖核验表存在", "「上游覆盖核验」章节无数据行（至少 1 行）");
    return;
  }

  /** 单元格是否为空理由：— / - / 空 / 裸 N/A。 */
  const isEmpty = (cell: string): boolean => cell.replace(/[—\-\sN/A/]/gi, "").length === 0;

  const problems: string[] = [];
  const pendingRows: string[] = [];
  dataRows.forEach((cells, i) => {
    const row = cells.join(" | ");
    // 残留 ❌ 待补 = 终稿硬伤
    if (row.includes("❌") || row.includes("待补")) {
      pendingRows.push(`行${i + 1}: ${row}`);
      return;
    }

    const issueCell = issueCol !== null && issueCol < cells.length ? cells[issueCol]! : "";
    const statusCell = statusCol !== null && statusCol < cells.length ? cells[statusCol]! : "";
    const reasonCell = reasonCol !== null && reasonCol < cells.length ? cells[reasonCol]! : "";

    // 全行搜 #issue 号
    const issueRefs = cells.join(" ").match(/#(\d+)/g) ?? [];
    const issueRefNums = issueRefs.map((r) => r.slice(1));
    const isNa = statusCell.includes("N/A") || issueCell.includes("N/A") || row.includes("N/A");

    if (issueRefNums.length === 0 && !isNa) {
      problems.push(`行${i + 1}: 既无 #issue 也无 N/A — ${row}`);
      return;
    }

    // N/A 行（无 issue）必须带实质理由
    if (isNa && issueRefNums.length === 0) {
      if (isEmpty(reasonCell)) {
        problems.push(`行${i + 1}: N/A 无理由（理由列须写一句话）— ${row}`);
      }
    }

    // 有 #issue 但指向不存在的编号 → 幽灵
    for (const ref of issueRefNums) {
      if (!realIssueIds.has(ref)) {
        problems.push(`行${i + 1}: #${ref} 不存在（幽灵引用）— ${row}`);
      }
    }
  });

  if (pendingRows.length > 0) {
    report.addFail(
      "覆盖核验表无待补残留",
      `${pendingRows.join("; ")}（终稿前必须转 ✅ 或 N/A）`,
    );
  } else if (problems.length > 0) {
    report.addFail("覆盖核验表形式", problems.join("; "));
  } else {
    report.addPass(
      "覆盖核验表形式",
      `${dataRows.length} 行，每行有 #issue 或 N/A+理由，无待补残留`,
    );
  }
}
